pkg/server/server.go

```diff
@@ -0,0 +1,184 @@
+	director := func(req *http.Request) {
+		token, err := tokenFromRequest(req)
+		if err != nil {
+			logger.Error(err, "token not present in request")
+			return
+		}
+		target, err := authz.TargetForToken(token)
+		if err != nil {
+			logger.Error(err, "could not get target for token")
+			return
+		}
+
+		targetQuery := target.RawQuery
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+		req.URL.Path, req.URL.RawPath = joinURLPath(target, req.URL)
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
+			req.Header.Set("User-Agent", "")
+		}
+	}
+
+	return &AzdoServer{
+		logger: logger.WithName("azdo-server"),
+		port:   port,
+		proxy:  &httputil.ReverseProxy{Director: director},
+		authz:  authz,
+	}
+}
+
+// ListenAndServe starts the HTTP server on the specified port
+func (a *AzdoServer) ListenAndServe(stopCh <-chan struct{}) {
+	a.logger.Info("Starting git proxy", "port", a.port)
+	router := mux.NewRouter()
+	router.HandleFunc("/readyz", readinessHandler(a.logger)).Methods("GET")
+	router.HandleFunc("/healthz", livenessHandler(a.logger)).Methods("GET")
+	router.PathPrefix("/").HandlerFunc(proxyHandler(a.logger, a.proxy, a.authz))
+	srv := &http.Server{Addr: fmt.Sprintf(":%s", a.port), Handler: router}
+
+	go func() {
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			a.logger.Error(err, "azdo-proxy server crashed")
+			os.Exit(1)
+		}
+	}()
+
+	<-stopCh
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if err := srv.Shutdown(ctx); err != nil {
+		a.logger.Error(err, "azdo-proxy server graceful shutdown failed")
+	} else {
+		a.logger.Info("azdo-proxy server stopped")
+	}
+}
+
+func proxyHandler(logger logr.Logger, p *httputil.ReverseProxy, authz auth.Authorization) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		token, err := tokenFromRequest(r)
+		if err != nil {
+			w.Header().Set("WWW-Authenticate", "Basic realm=\"Restricted\"")
+			http.Error(w, "Missing basic authentication", http.StatusUnauthorized)
+			return
+		}
+
+		// Check basic auth with local auth configuration
+		err = authz.IsPermitted(r.URL.EscapedPath(), token)
+		if err != nil {
+			logger.Error(err, "Received unauthorized request")
+			http.Error(w, "User not permitted", http.StatusForbidden)
+			return
+		}
+		pat, err := authz.TargetForToken(token)
+
+		// Forward request to destination server
+		logger.Info("Authenticated request", "path", r.URL.Path)
+		r.Header.Del("Authorization")
+		patB64 := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("pat:%s", pat)))
+		r.Header.Add("Authorization", "Basic "+patB64)
```

Comment:
This could be a small descriptive function, like `setAzureDevOpsPAT()`

Reply:
Why?